code/espurna/flow.ino

```diff
@@ -0,0 +1,782 @@
+                    int cmp = s1 == NULL ? (s2 == NULL ? 0 : -1) :
+                              s2 == NULL ? 1 :
+                              strcmp(s1, s2);
+                    r = _operation.equals("=") ? cmp == 0 :
+                        _operation.equals(">") ? cmp > 0 :
+                        /*_operation.equals("<") ?*/ cmp < 0
+                    ;
+                } else if (_data->is<bool>()) {
+                    bool b1 = _data->as<bool>();
+                    bool b2 = _test->as<bool>();
+                    r = _operation.equals("=") ? b1 == b2 :
+                        _operation.equals(">") ? b1 > b2 :
+                        /*_operation.equals("<") ?*/ b1 < b2
+                    ;
+                }
+                processOutput(*_data, r ? 0 : 1);
+            }
+        }
+
+        static void reg() {
+            flowRegisterComponent("Compare", &flow_compare_component, flow_compare_component_json,
+                (flow_component_factory_f)([] (JsonObject& properties) { return new FlowCompareComponent(properties); }));
+        }
+};
+
+// -----------------------------------------------------------------------------
+// Delay component
+// -----------------------------------------------------------------------------
+
+PROGMEM const char flow_reset[] = "Reset";
+PROGMEM const char* const flow_delay_inputs[] = {flow_data, flow_reset};
+
+PROGMEM const FlowConnections flow_delay_component = {
+    2, flow_delay_inputs,
+    1, flow_data_array,
+};
+
+PROGMEM const char flow_delay_component_json[] =
+    "\"Delay\": "
+    "{"
+        "\"name\":\"Delay\","
+        "\"icon\":\"pause\","
+        "\"inports\":[{\"name\":\"Data\",\"type\":\"any\"},{\"name\":\"Reset\",\"type\":\"any\"}],"
+        "\"outports\":[{\"name\":\"Data\",\"type\":\"any\"}],"
+        "\"properties\":[{\"name\":\"Seconds\",\"type\":\"int\"}, {\"name\":\"Last only\",\"type\":\"bool\"}]"
+    "}";
+
+class FlowDelayComponent : public FlowComponent {
+    private:
+        struct scheduled_task_t {
+            FlowDelayComponent *component;
+            Ticker *ticker;
+            JsonVariant *data;
+        };
+
+        long _time;
+        bool _lastOnly;
+        int _queueSize = 0;
+        long _skipUntil = 0;
+
+    public:
+        FlowDelayComponent(JsonObject& properties) {
+            _time = 1000 * (int)properties["Seconds"];
+            _lastOnly = properties["Last only"];
+        }
+
+        virtual void processInput(JsonVariant& data, int inputNumber) {
+            if (inputNumber == 0) { // data
+                Ticker *ticker = new Ticker();
+                scheduled_task_t *task = new flow_scheduled_task_t();
+                task->component = this;
+                task->ticker = ticker;
+                task->data = clone(data);
+
+                ticker->once_ms(_time, onDelay, task);
+
+                _queueSize++;
+            } else { // reset
+                _skipUntil = millis() + _time;
```

Comment:
Essentially, timers should calculate relative time.
`now_millis - recorded_millis > interval`
And am I understanding this will create ticker per input? Can it use only a single instance? I'll cancel old timer and start a new one

btw Ticker can only work with maximum value of 1h40m (tried some time ago to reimplement scheduler using it: https://github.com/xoseperez/espurna/issues/678#issuecomment-395197349)

Reply:
> And am I understanding this will create ticker per input?

Right, ticker per input, because there could be several inputs waiting in the queue and one ticker can handle only one timer.

> Essentially, timers should calculate relative time.

I was going to slightly rework that code to do not use time comparison

Reply:
I see. Note that ticker function triggers in a system context, so does everything else in a process chain after that point. And with weird timing restriction, is it worth using it at all like that?
*there are no problems right now, but using it inside loop() you can allow yourself to go back into the system via yield() / delay(), but not other way around*

esp8266 arduino even has a nice abstraction for comparisons:
https://github.com/esp8266/Arduino/blob/master/cores/esp8266/PolledTimeout.h

Reply:
Amendment: new Ticker implements a special mechanism to execute function in a loop() context (aka CONT) by using schedule_function() (which I totally knew about...). see:
https://github.com/esp8266/Arduino/blob/64e30b270bcb8accde0c62d5998fd69907d764b6/libraries/Ticker/Ticker.h#L46
[Schedule.h](https://github.com/esp8266/Arduino/blob/master/cores/esp8266/Schedule.h), [.cpp](https://github.com/esp8266/Arduino/blob/master/cores/esp8266/Schedule.cpp)

Note:
the hardcoded limit of 32 functions.
this is not available in the 2.3.0
...or replicate tem, place our handler at the end of the loop()